templates/agent/scripts/couchbase.py.j2

```diff
@@ -0,0 +1,158 @@
+    # then take average of values in this list.
+    if data_type == "avg":
+        assert type(value) == list
+        return sum(value) / float(len(value))
+
+    # value should be a list of string
+    if data_type == "string_list":
+        assert type(value) == list
+        return json.dumps({
+            "data": [{"%s%s%s" % ("{", macro, "}") : str(item)} for item in value]
+        })
+
+def get_data_from_keys(data, keys_string):
+    """
+    Get wanted data through layers.
+    """
+    keys = keys_string.split(".")
+    for key in keys:
+        data = data[key]
+    return data
+
+
+def handle_cluster(args):
+    # couchbase['cluster', key, type]
+    data = get_from_url(cluster_url)
+    data_type = args[1]
+    keys_string = args[0]
+
+    result = get_data_from_keys(data, keys_string)
+
+    if keys_string == "nodes":
+        result = [item['hostname'].split(":")[0] for item in result]
+        print handle_data_by_type(result, data_type, "#NODEHOST")
+        exit(0)
+
+    print handle_data_by_type(result, data_type)
+
+
+def handle_bucket(args):
+    # warning: you can not call your bucket name to "names",
+    # it will conflict with the following condition.
+    # couchbase['bucket', 'names', type]
+    # type here is `string_list`
+    if len(args) == 2 and args[0] == 'names':
+        data = get_from_url(buckets_url)
+        bucket_names = [bucket['name'] for bucket in data]
+        print handle_data_by_type(bucket_names, args[1], "#BUCKETNAME")
+        exit(0)
+
+    # couchbase['bucket', name, key, type]
+    name = args[0]
+    data_type = args[2]
+
+    bucket_url = urlparse.urljoin(buckets_url, "%s/" % name)
+
+    # couchbase['bucket', name, 'stats.*', type]
+    keys_string = args[1]
+    if keys_string.startswith("stats"):
+        bucket_stats_url = urlparse.urljoin(bucket_url, "stats?zoom=minute")
+        data = get_from_url(bucket_stats_url)
+        result = get_data_from_keys(data, keys_string.split(".", 1)[1])
+    else:
+        data = get_from_url(bucket_url)
+        # couchbase['bucket', name, "nodes.hostnames", "string"]
+        if keys_string == "nodes.hostnames":
+            result = [node['hostname'] for node in get_data_from_keys(data, "nodes")]
+            result = ", ".join(result)
+        # couchbase['bucket', name, "*", type]
+        else:
+            result = get_data_from_keys(data, keys_string)
+
+    print handle_data_by_type(result, data_type)
+
+
+def handle_node(args):
+    # couchbase['node', hostname, key, type]
+    hostname = args[0]
+    nodes = get_from_url(cluster_url).get("nodes")
+    data_type = args[2]
+    for node in nodes:
+        if node.get("hostname").split(":")[0] == hostname:
+            data = node
+            break
+    result = get_data_from_keys(data, args[1])
+
+    print handle_data_by_type(result, data_type)
+
+
+def main():
+    args = sys.argv[1:]
+
+    if len(args) < 1:
+        print "argument not provided."
+        exit(1)
```

Comment:
Better raise exception with error messages and capture them than managing the script exit manually